File: src/diffvouch/review.py
from __future__ import annotations

from pathlib import Path

from diffvouch.config import RepositoryConfig, load_repository_config
from diffvouch.errors import ArgumentError
from diffvouch.git import chunk_patch, collect_diff, repository_root, run_git
from diffvouch.models import (
    Dimensions,
    FilesSummary,
    ProviderInfo,
    ProviderReview,
    Publication,
    ReviewResult,
    Scope,
    Severity,
)
from diffvouch.prompt import build_review_prompt
from diffvouch.providers import Provider, create_provider
from diffvouch.rating import add_finding_ids, calculate_rating, evaluate_gate
from diffvouch.sanitize import redact_secrets


def _combine(reviews: list[ProviderReview], sizes: list[int]) -> ProviderReview:
    if len(reviews) == 1:
        return reviews[0]
    total = sum(sizes)
    dimension_values: dict[str, float] = {}
    for name in Dimensions.model_fields:
        dimension_values[name] = round(
            sum(getattr(review.dimensions, name) * size for review, size in zip(reviews, sizes))
            / total,
            2,
        )
    findings = []
    seen: set[tuple[str | None, int | None, str]] = set()
    for review in reviews:
        for finding in review.findings:
            identity = (finding.path, finding.line, finding.title.casefold())
            if identity not in seen:
                seen.add(identity)
                findings.append(finding)
    positives = list(
        dict.fromkeys(item for review in reviews for item in review.positive_observations)
    )
    verification = list(
        dict.fromkeys(item for review in reviews for item in review.needs_verification)
    )
    summaries = " ".join(review.summary.strip() for review in reviews)
    return ProviderReview(
        summary=f"Reviewed the complete patch in {len(reviews)} chunks. {summaries}",
        dimensions=Dimensions(**dimension_values),
        findings=findings,
        positive_observations=positives,
        needs_verification=verification,
    )


def perform_review(
    *,
    provider_name: str,
    transport: str | None,
    model: str | None,
    effort: str | None,
    base: str | None,
    committed_only: bool,
    staged_only: bool,
    excludes: list[str],
    config_path: str | None,
    max_diff_bytes: int | None,
    fail_below: float | None,
    fail_on_severity: str | None,
    publication_requested: bool,
    root: Path | None = None,
) -> ReviewResult | None:
    repository = repository_root(root)
    trusted_ref = (
        run_git(repository, "rev-parse", "--verify", f"{base}^{{commit}}").strip()
        if base
        else "HEAD"
    )
    config: RepositoryConfig = load_repository_config(
        repository, config_path, trusted_ref=trusted_ref
    )
    selected_transport = transport or "cli"
    selected_model = model or getattr(config.provider.models, provider_name)
    patterns = [*config.review.exclude, *excludes]
    collected = collect_diff(
        root=repository,
        base=base,
        committed_only=committed_only,
        staged_only=staged_only,
        excludes=patterns,
        max_diff_bytes=max_diff_bytes or config.review.max_diff_bytes,
    )
    if not collected.patch:
        return None
    sanitized, redaction_count = redact_secrets(collected.patch)
    chunks = chunk_patch(sanitized, config.review.chunk_bytes)
    provider: Provider = create_provider(provider_name, selected_transport, selected_model, effort)
    reviews: list[ProviderReview] = []
    sizes: list[int] = []
    for index, chunk in enumerate(chunks, start=1):
        prompt = build_review_prompt(
            patch=chunk,
            chunk_index=index,
            chunk_count=len(chunks),
            rubric=config.review.rubric,
            instructions=config.review.instructions,
        )
        reviews.append(provider.review(prompt))
        sizes.append(len(chunk.encode()))
    provider_review = _combine(reviews, sizes)

    reviewed_paths = set(collected.reviewed_files)
    accepted = []
    needs_verification = list(provider_review.needs_verification)
    for finding in provider_review.findings:
        if finding.path is not None and finding.path not in reviewed_paths:
            needs_verification.append(
                f"Provider cited {finding.path!r}, which was not part of the reviewed patch."
            )
            continue
        accepted.append(finding)
    provider_review.findings = accepted
    provider_review.needs_verification = needs_verification
    findings = add_finding_ids(provider_review)
    rating = calculate_rating(provider_review, config.review.rubric)
    selected_fail_below = fail_below if fail_below is not None else config.quality_gate.fail_below
    selected_fail_severity = fail_on_severity or config.quality_gate.fail_on_severity
    try:
        severity = Severity(selected_fail_severity) if selected_fail_severity else None
    except ValueError as exc:
        raise ArgumentError(f"invalid fail-on severity {selected_fail_severity!r}") from exc
    gate = evaluate_gate(rating, findings, selected_fail_below, severity)
    return ReviewResult(
        scope=Scope(
            mode=collected.mode,
            base_ref=collected.base_ref,
            base_sha=collected.base_sha,
            merge_base=collected.merge_base,
            head_sha=collected.head_sha,
        ),
        provider=ProviderInfo(
            name=provider_name,
            transport=selected_transport,
            model=provider.model,
            effort=effort,
        ),
        rating=rating,
        summary=provider_review.summary,
        findings=findings,
        positive_observations=provider_review.positive_observations,
        needs_verification=provider_review.needs_verification,
        files=FilesSummary(
            reviewed=collected.reviewed_files,
            excluded=collected.excluded_files,
            binary=collected.binary_files,
            omitted=[],
            redactions=redaction_count,
        ),
        gate=gate,
        publication=Publication(requested=publication_requested),
    )

File: internal/review/review.go
package review

import (
	"fmt"
	"math"
	"sort"
	"strings"

	"github.com/divyangchauhan/DiffVouch/internal/config"
	"github.com/divyangchauhan/DiffVouch/internal/dv"
	"github.com/divyangchauhan/DiffVouch/internal/gitdiff"
	"github.com/divyangchauhan/DiffVouch/internal/github"
	"github.com/divyangchauhan/DiffVouch/internal/model"
	"github.com/divyangchauhan/DiffVouch/internal/provider"
	"github.com/divyangchauhan/DiffVouch/internal/rating"
	"github.com/divyangchauhan/DiffVouch/internal/sanitize"
)

type Options struct {
	ProviderName         string
	Transport            string
	Model                string
	Effort               string
	Base                 string
	CommittedOnly        bool
	StagedOnly           bool
	Excludes             []string
	ConfigPath           string
	MaxDiffBytes         int
	FailBelow            *float64
	FailOnSeverity       string
	PublicationRequested bool
	ReviewPrompt         string
	Root                 string
}

func Perform(options Options) (*model.ReviewResult, *model.FilesSummary, error) {
	repo, err := gitdiff.RepositoryRoot(options.Root)
	if err != nil {
		return nil, nil, err
	}
	trustedRef := "HEAD"
	if options.Base != "" {
		resolved, resolveErr := gitdiff.Run(repo, 4096, "rev-parse", "--verify", options.Base+"^{commit}")
		if resolveErr != nil {
			return nil, nil, resolveErr
		}
		trustedRef = strings.TrimSpace(resolved)
	}
	repositoryConfig, err := config.LoadRepository(repo, options.ConfigPath, trustedRef)
	if err != nil {
		return nil, nil, dv.Wrap(dv.ExitArguments, "load repository configuration", err)
	}
	transport := options.Transport
	if transport == "" {
		transport = "cli"
	}
	selectedModel := options.Model
	if selectedModel == "" {
		if options.ProviderName == "codex" {
			selectedModel = repositoryConfig.Provider.Models.Codex
		} else {
			selectedModel = repositoryConfig.Provider.Models.Claude
		}
	}
	maxBytes := options.MaxDiffBytes
	if maxBytes == 0 {
		maxBytes = repositoryConfig.Review.MaxDiffBytes
	}
	collected, err := gitdiff.Collect(gitdiff.Options{
		Root: repo, Base: options.Base, CommittedOnly: options.CommittedOnly,
		StagedOnly:   options.StagedOnly,
		Excludes:     append(append([]string{}, repositoryConfig.Review.Exclude...), options.Excludes...),
		MaxDiffBytes: maxBytes,
	})
	if err != nil {
		return nil, nil, err
	}
	if collected.Patch == "" {
		files := &model.FilesSummary{Reviewed: collected.ReviewedFiles, Excluded: collected.ExcludedFiles, Binary: collected.BinaryFiles, Omitted: []string{}}
		return nil, files, nil
	}
	redacted, redactions := sanitize.Redact(collected.Patch)
	chunks, err := gitdiff.ChunkPatch(redacted, repositoryConfig.Review.ChunkBytes)
	if err != nil {
		return nil, nil, err
	}
	adapter, err := provider.New(provider.Options{Name: options.ProviderName, Transport: transport, Model: selectedModel, Effort: options.Effort})
	if err != nil {
		return nil, nil, err
	}
	reviews := make([]model.ProviderReview, 0, len(chunks))
	sizes := make([]int, 0, len(chunks))
	for index, chunk := range chunks {
		prompt := buildPrompt(chunk, index+1, len(chunks), repositoryConfig.Review.Rubric, repositoryConfig.Review.Instructions, options.ReviewPrompt)
		providerReview, reviewErr := adapter.Review(prompt)
		if reviewErr != nil {
			return nil, nil, reviewErr
		}
		reviews = append(reviews, providerReview)
		sizes = append(sizes, len([]byte(chunk)))
	}
	combined := combine(reviews, sizes)
	reviewedPaths := map[string]struct{}{}
	for _, path := range collected.ReviewedFiles {
		reviewedPaths[path] = struct{}{}
	}
	accepted, rejectedLocations := validateFindingLocations(combined.Findings, reviewedPaths, github.ChangedLines(collected.Patch))
	combined.NeedsVerification = append(combined.NeedsVerification, rejectedLocations...)
	combined.Findings = rating.NormalizeFindings(accepted)
	if combined.Findings == nil {
		combined.Findings = []model.Finding{}
	}
	if combined.PositiveObservations == nil {
		combined.PositiveObservations = []string{}
	}
	if combined.NeedsVerification == nil {
		combined.NeedsVerification = []string{}
	}
	calculatedRating := rating.Calculate(combined, repositoryConfig.Review.Rubric)
	failBelow := options.FailBelow
	if failBelow == nil {
		failBelow = repositoryConfig.QualityGate.FailBelow
	}
	failSeverity := options.FailOnSeverity
	if failSeverity == "" {
		failSeverity = repositoryConfig.QualityGate.FailOnSeverity
	}
	if failSeverity != "" && !rating.ValidSeverity(failSeverity) {
		return nil, nil, dv.New(dv.ExitArguments, "invalid fail-on severity")
	}
	var effort *string
	if options.Effort != "" {
		effort = &options.Effort
	}
	result := &model.ReviewResult{
		SchemaVersion: 1, ReviewID: model.NewReviewID(), Status: "complete", Partial: false,
		Scope:    model.Scope{Mode: collected.Mode, BaseRef: collected.BaseRef, BaseSHA: collected.BaseSHA, MergeBase: collected.MergeBase, HeadSHA: collected.HeadSHA},
		Provider: model.ProviderInfo{Name: options.ProviderName, Transport: transport, Model: adapter.Model(), Effort: effort},
		Rating:   calculatedRating, Summary: combined.Summary, Findings: combined.Findings,
		PositiveObservations: combined.PositiveObservations, NeedsVerification: combined.NeedsVerification,
		Files:       model.FilesSummary{Reviewed: collected.ReviewedFiles, Excluded: collected.ExcludedFiles, Binary: collected.BinaryFiles, Omitted: []string{}, Redactions: redactions},
		Gate:        rating.Gate(calculatedRating, combined.Findings, failBelow, failSeverity),
		Publication: model.Publication{Requested: options.PublicationRequested},
	}
	return result, nil, nil
}

func validateFindingLocations(findings []model.Finding, reviewedPaths map[string]struct{}, changedLines map[github.DiffLocation]struct{}) ([]model.Finding, []string) {
	accepted := make([]model.Finding, 0, len(findings))
	var needsVerification []string
	for _, finding := range findings {
		if finding.Path != nil {
			if _, ok := reviewedPaths[*finding.Path]; !ok {
				needsVerification = append(needsVerification, fmt.Sprintf("Provider cited %q, which was not part of the reviewed patch.", *finding.Path))
				continue
			}
		}
		if finding.Line != nil && finding.Path != nil && finding.Side != nil {
			side := "RIGHT"
			if *finding.Side == "old" {
				side = "LEFT"
			}
			if _, ok := changedLines[github.DiffLocation{Path: *finding.Path, Side: side, Line: *finding.Line}]; !ok {
				needsVerification = append(needsVerification, fmt.Sprintf("Provider cited %s:%d on an unchanged or unavailable line.", *finding.Path, *finding.Line))
				continue
			}
		}
		accepted = append(accepted, finding)
	}
	return accepted, needsVerification
}

func combine(reviews []model.ProviderReview, sizes []int) model.ProviderReview {
	if len(reviews) == 1 {
		return reviews[0]
	}
	total := 0
	for _, size := range sizes {
		total += size
	}
	combined := model.ProviderReview{}
	var correctness, security, maintainability, testing, scope float64
	seenFindings := map[string]struct{}{}
	seenPositive := map[string]struct{}{}
	seenVerification := map[string]struct{}{}
	var summaries []string
	for index, review := range reviews {
		weight := float64(sizes[index]) / float64(total)
		correctness += review.Dimensions.Correctness * weight
		security += review.Dimensions.Security * weight
		maintainability += review.Dimensions.Maintainability * weight
		testing += review.Dimensions.Testing * weight
		scope += review.Dimensions.Scope * weight
		summaries = append(summaries, strings.TrimSpace(review.Summary))
		for _, finding := range review.Findings {
			path := ""
			line := 0
			if finding.Path != nil {
				path = *finding.Path
			}
			if finding.Line != nil {
				line = *finding.Line
			}
			key := fmt.Sprintf("%s:%d:%s", path, line, strings.ToLower(finding.Title))
			if _, ok := seenFindings[key]; !ok {
				seenFindings[key] = struct{}{}
				combined.Findings = append(combined.Findings, finding)
			}
		}
		for _, value := range review.PositiveObservations {
			if _, ok := seenPositive[value]; !ok {
				seenPositive[value] = struct{}{}
				combined.PositiveObservations = append(combined.PositiveObservations, value)
			}
		}
		for _, value := range review.NeedsVerification {
			if _, ok := seenVerification[value]; !ok {
				seenVerification[value] = struct{}{}
				combined.NeedsVerification = append(combined.NeedsVerification, value)
			}
		}
	}
	round := func(value float64) float64 { return math.Round(value*100) / 100 }
	combined.Dimensions = model.Dimensions{Correctness: round(correctness), Security: round(security), Maintainability: round(maintainability), Testing: round(testing), Scope: round(scope)}
	combined.Summary = fmt.Sprintf("Reviewed the complete patch in %d chunks. %s", len(reviews), strings.Join(summaries, " "))
	sort.Strings(combined.PositiveObservations)
	sort.Strings(combined.NeedsVerification)
	return combined
}
